src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+            {
+                return Interop.Kernel32.GetProcAddress(libHandle, nameof(Interop.Winsock.GetAddrInfoExCancel)) != IntPtr.Zero;
+            }
+        }
+
+        public static unsafe void GetAddrInfoAsync(string name, DnsResolveAsyncResult asyncResult)
+        {
+            try
+            { }
+            finally
+            {
+                GetAddrInfoExContext* context = GetAddrInfoExContext.AllocateContext(name, asyncResult);
+
+                AddressInfoEx hints = new AddressInfoEx();
+                hints.ai_flags = AddressInfoHints.AI_CANONNAME;
+                hints.ai_family = AddressFamily.Unspecified; // Gets all address families
+
+                SocketError errorCode =
+                    (SocketError)Interop.Winsock.GetAddrInfoExW(name, null, 0 /* NS_ALL*/, IntPtr.Zero, ref hints, out context->Result, IntPtr.Zero, ref context->Overlapped, s_getAddrInfoExCallback, out context->CancelHandle);
+
+                if (errorCode != SocketError.IOPending)
+                    ProcessResult(errorCode, context);
+            }
+        }
+
+        private static unsafe void GetAddressInfoExCallback([In] int error, [In] int bytes, [In] NativeOverlapped* overlapped)
+        {
+            try
+            { }
+            finally
+            {
+                // Can be casted directly to QueryContext* because the overlapped is its first field
+                GetAddrInfoExContext* context = (GetAddrInfoExContext*)overlapped;
+
+                ProcessResult((SocketError)error, context);
+            }
+        }
+
+        private static unsafe void ProcessResult(SocketError errorCode, GetAddrInfoExContext* context)
+        {
+            GetAddrInfoExState state = context->GetQueryState();
+
+            if (state == null || !state.SetCallbackStartedOrCanceled())
+                return;
+
+            try
+            {
+                if (errorCode != SocketError.Success)
+                {
+                    state.CompleteAsyncResult(new SocketException((int)errorCode));
+                    return;
+                }
+
+                AddressInfoEx* result = context->Result;
+                string canonicalName = null;
+
+                List<IPAddress> addresses = new List<IPAddress>();
+
+                while (result != null)
+                {
+                    if (canonicalName == null && result->ai_canonname != IntPtr.Zero)
+                        canonicalName = Marshal.PtrToStringUni(result->ai_canonname);
+
+                    IPAddress ipAddress = null;
+
+                    if (result->ai_family == AddressFamily.InterNetwork)
+                    {
+                        ipAddress = CreateIPv4Address(result->ai_addr, result->ai_addrlen);
+                    }
+                    else if (SocketProtocolSupportPal.OSSupportsIPv6 && result->ai_family == AddressFamily.InterNetworkV6)
```

Comment:
Will we ever get to this point if OSSupportsIPv6 is false?

Reply:
Discussed above, I will leave the check here in case there is a code change in DNS.cs to call this method even if OS doesn't support IPv6